torch/csrc/toffee/export.cpp

```diff
@@ -0,0 +1,302 @@
+#include "torch/csrc/toffee/export.h"
+#include "torch/csrc/autograd/primspec.h"
+#include "torch/csrc/utils/python_numbers.h"
+#include "torch/csrc/utils/python_strings.h"
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/toffee.h"
+
+#include "torch/csrc/autograd/functions/convolution.h"
+#include "torch/csrc/jit/dead_code_elimination.h"
+#include "torch/csrc/utils/functional.h"
+#include <ATen/ATen.h>
+
+#include <fstream>
+#undef NDEBUG
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#define NDEBUG
+
+namespace py = pybind11;
+
+namespace torch { namespace jit {
+
+std::string node_name(Node* n) {
+  return n->uniqueName();
+}
+
+// transform PythonOps and Cpp Ops into Node's that match ToffeeIR
+// semantics.
+// Eventually this should just be part of init_pass but we should avoid
+// tight coupling of the JIT and Toffee IR exporter until ready.
+std::shared_ptr<Graph> ToToffeeIR(std::shared_ptr<Graph>& g,
+                                  const std::unordered_map<void*, Node*>& old_buffer_map) {
+  torch::autograd::PrimSpecContext ctx;
+  std::unordered_map<Node*, Node*> env;
+  std::shared_ptr<Graph> out_graph = std::make_shared<Graph>();
+  ctx.graph = out_graph.get();
+  for (auto input : g->inputs())
+    env[input] = ctx.graph->addInput()->setType(input->typeOption());
+  auto envFn = [&env](Node * n) {
+    auto it = env.find(n);
+    JIT_ASSERTM(it != env.end(), "Dangling node reference");
+    JIT_ASSERTM(it->second, "Unused node was subsequently used");
+    return it->second;
+  };
+  std::unordered_map<void*, Node*> buffer_map;
+  for (auto kv : old_buffer_map) {
+    buffer_map[kv.first] = envFn(kv.second);
+  }
+  ctx.buffer_map = &buffer_map;
+  // put the new outputs in our environment map, and
+  // copy the type from the input graph if they were not set by the
+  // primspec
+  auto setOutputs = [&](Node * node, const node_list & outputs) {
+    auto old_outputs = node->outputs();
+    JIT_ASSERTM(outputs.size() <= old_outputs.size(), "primspec produced too many outputs");
+    size_t i = 0;
+    for(auto & old : old_outputs) {
+      // TODO: what if there are multiple trailing handle outputs?  That is
+      // a serious invariant violation...
+      if(i >= outputs.size()) {
+        // primspecs do not deal with Handles at the moment
+        // so we map handles to Unused nodes
```

Comment:
Except that Unused nodes no longer exist, so it would be good to change this comment